### PR TITLE
Implement constexpr algorithms.

### PR DESCRIPTION
**Pull request by BillyONeal**
Resolves GH-6 ( P0202R3 ), GH-38 ( P0879R0 ), and drive-by fixes GH-414.

Constexprizes the following algorithms by adding constexpr, _CONSTEXPR20, and _CONSTEXPR20_ICE to things:

* adjacent_find
* all_of
* any_of
* binary_search
* copy
* copy_backward
* copy_if
* copy_n
* count
* count_if
* equal
* equal_range
* exchange
* fill
* fill_n
* find
* find_end
* find_first_of
* find_if
* find_if_not
* for_each
* for_each_n
* generate
* generate_n
* includes
* is_heap
* is_heap
* is_heap_until
* is_partitioned
* is_permutation
* is_sorted
* is_sorted_until
* iter_swap
* lexicographical_compare
* lower_bound
* make_heap
* merge
* mismatch
* move
* move_backward
* next_permutation
* none_of
* nth_element
* partial_sort
* partial_sort_copy
* partition
* partition_copy
* partition_point
* pop_heap
* prev_permutation
* push_heap
* remove
* remove_copy
* remove_copy_if
* remove_if
* replace
* replace_copy
* replace_copy_if
* replace_if
* reverse_copy
* revese
* rotate
* rotate_copy
* search
* search_n
* set_difference
* set_intersection
* set_symmetric_difference
* set_union
* sort
* sort_heap
* swap
* swap_ranges
* transform
* unique
* unique_copy
* upper_bound

Specific notes:

`skipped_tests.txt`: Turn on all tests previously blocked by missing constexpr algorithms (and `exchange` and `swap`). Mark those algorithms that cannot be turned on that we have outstanding PRs for with their associated PRs.
`yvals_core.h`: Turn on feature test macros.
`xutility`: Move the _Ptr_cat family down to copy, and fix associated SHOUTY comments to indicate that this is really an implementation detail of copy, not something the rest of the standard library intends to use directly. Removed and clarified some of the comments as requested by Casey Carter.

Updates the llvm submodule to get https://github.com/llvm/llvm-project/commit/6d8abe424a77f736fbed114eeac574b9bfe6b0c1

# Checklist

Be sure you've read README.md and understand the scope of this repo.

If you're unsure about a box, leave it unchecked. A maintainer will help you.

- [x] Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.
- [x] The STL builds successfully and all tests have passed (must be manually
  verified by an STL maintainer before automated testing is enabled on GitHub,
  leave this unchecked for initial submission).
- [x] These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).
- [x] These changes were written from scratch using only this repository,
  the C++ Working Draft (including any cited standards), other WG21 papers
  (excluding reference implementations outside of proposed standard wording),
  and LWG issues as reference material. If they were derived from a project
  that's already listed in NOTICE.txt, that's fine, but please mention it.
  If they were derived from any other project (including Boost and libc++,
  which are not yet listed in NOTICE.txt), you *must* mention it here,
  so we can determine whether the license is compatible and what else needs
  to be done.

### CODE REVIEW

**Comment by BillyONeal** [comment]
FYI I haven't submitted an internal PR yet because I'm still working on adding tests, but wanted GH contributors to be able to see this when it was done.

File: stl/inc/xutility [hunk 1]
```diff
@@ -3662,27 +3674,40 @@ _DestTy* copy(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Last, _DestTy (&_Dest)[_DestSize
...
 
 // FUNCTION TEMPLATE copy_n
+template <class _InIt, class _Diff, class _OutIt>
+constexpr _OutIt _Copy_n_core(_InIt _First, _Diff _Count, _OutIt _Dest) {
+    for (;;) {
+        *_Dest = *_First;
+        ++_Dest;
+        --_Count;
+        if (_Count == 0) { // note that we avoid an extra ++_First here to allow istream_iterator to work,
+                           // see LWG-2471
```
Comment (CaseyCarter) [inline stl/inc/xutility]: **LWG-2471 RAGE INTENSIFIES**

I can't get over how broken this is: `copy_n(i, 0, r)` needs to force `*i` back into the stream to be consistent with `copy_n(i, 1, r)`. 

LWG-2242 and LWG-2471 are the "`std::copy_n` is a horrible mess that we can't fix without breaking existing code" issues.

(No change requested.)

File: stl/inc/xutility [hunk 2]
```diff
@@ -3662,27 +3674,40 @@ _DestTy* copy(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Last, _DestTy (&_Dest)[_DestSize
 #endif // _HAS_CXX17
 
 // FUNCTION TEMPLATE copy_n
+template <class _InIt, class _Diff, class _OutIt>
+constexpr _OutIt _Copy_n_core(_InIt _First, _Diff _Count, _OutIt _Dest) {
```
Comment (CaseyCarter) [inline stl/inc/xutility]: `_core` seems really weird to me. Why not `_Copy_n_impl`?

File: stl/inc/xutility [hunk 3]
```diff
@@ -3585,11 +3585,22 @@ _OutIt _Copy_memmove(move_iterator<_InIt> _First, move_iterator<_InIt> _Last, _O
 
 #if _HAS_IF_CONSTEXPR
 template <class _InIt, class _OutIt>
-_OutIt _Copy_unchecked(_InIt _First, _InIt _Last, _OutIt _Dest) {
+_CONSTEXPR20_ICE _OutIt _Copy_unchecked(_InIt _First, _InIt _Last, _OutIt _Dest) {
     // copy [_First, _Last) to [_Dest, ...)
     // note: _Copy_unchecked is called directly from elsewhere in the STL
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/xutility]: Ditto 3915 and 4006.

File: stl/inc/algorithm [hunk 4]
```diff
@@ -243,7 +245,8 @@ _NODISCARD pair<_InIt1, _InIt2> mismatch(_InIt1 _First1, const _InIt1 _Last1, _I
 
 #if _ITERATOR_DEBUG_ARRAY_OVERLOADS
 template <class _InIt1, class _RightTy, size_t _RightSize, class _Pr, enable_if_t<!is_same_v<_RightTy*, _Pr>, int> = 0>
-_NODISCARD pair<_InIt1, _RightTy*> mismatch(const _InIt1 _First1, const _InIt1 _Last1, _RightTy (&_First2)[_RightSize],
+_NODISCARD _CONSTEXPR20 pair<_InIt1, _RightTy*> mismatch(const _InIt1 _First1, const _InIt1 _Last1,
+    _RightTy (&_First2)[_RightSize],
     _Pr _Pred) { // return [_First1, _Last1)/[_First2, ...) mismatch using _Pred
```
Comment (CaseyCarter) [inline stl/inc/algorithm]: Wrap comment?

File: stl/inc/algorithm [hunk 5]
```diff
@@ -665,13 +673,27 @@ _NODISCARD _FwdIt partition_point(_FwdIt _First, _FwdIt _Last, _Pr _Pred) {
...
+
+            return true;
+        } else
+#endif // __cpp_lib_is_constant_evaluated
+        {
+            const auto _First1_ch = reinterpret_cast<const char*>(_First1);
+            const auto _First2_ch = reinterpret_cast<const char*>(_First2);
+            const auto _Count     = static_cast<size_t>(reinterpret_cast<const char*>(_Last2) - _First2_ch);
+            return _CSTD memcmp(_First1_ch, _First2_ch, _Count) == 0;
+        }
```
Comment (CaseyCarter) [inline stl/inc/algorithm]: Can't we do better with `__builtin_memcmp`? At least for non-CUDA?

File: stl/inc/algorithm [hunk 6]
```diff
@@ -1359,20 +1399,21 @@ _OutIt transform(const _InIt1 _First1, const _InIt1 _Last1, const _InIt2 _First2
 
 #if _ITERATOR_DEBUG_ARRAY_OVERLOADS
 template <class _InIt1, class _RightTy, size_t _RightSize, class _OutIt, class _Fn>
-_OutIt transform(const _InIt1 _First1, const _InIt1 _Last1, _RightTy (&_First2)[_RightSize], const _OutIt _Dest,
+_CONSTEXPR20 _OutIt transform(const _InIt1 _First1, const _InIt1 _Last1, _RightTy (&_First2)[_RightSize],
+    const _OutIt _Dest,
     _Fn _Func) { // transform [_First1, _Last1) and [_First2, ...), array source
```
Comment (CaseyCarter) [inline stl/inc/algorithm]: Wrap comment.

File: stl/inc/algorithm [hunk 7]
```diff
@@ -1583,7 +1626,8 @@ void generate(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Fn _Func) noexcept /* termi
 
 // FUNCTION TEMPLATE generate_n
 template <class _OutIt, class _Diff, class _Fn>
-_OutIt generate_n(_OutIt _Dest, const _Diff _Count_raw, _Fn _Func) { // replace [_Dest, _Dest + _Count) with _Func()
+_CONSTEXPR20 _OutIt generate_n(
+    _OutIt _Dest, const _Diff _Count_raw, _Fn _Func) { // replace [_Dest, _Dest + _Count) with _Func()
```
Comment (CaseyCarter) [inline stl/inc/algorithm]: Wrap comment.

File: stl/inc/algorithm [hunk 8]
```diff
@@ -1636,7 +1680,8 @@ _DestTy* generate_n(_ExPo&&, _DestTy (&_Dest)[_DestSize], const _Diff _Count_raw
 
 // FUNCTION TEMPLATE remove_copy
 template <class _InIt, class _OutIt, class _Ty>
-_OutIt remove_copy(_InIt _First, _InIt _Last, _OutIt _Dest, const _Ty& _Val) { // copy omitting each matching _Val
+_CONSTEXPR20 _OutIt remove_copy(
+    _InIt _First, _InIt _Last, _OutIt _Dest, const _Ty& _Val) { // copy omitting each matching _Val
```
Comment (CaseyCarter) [inline stl/inc/algorithm]: Wrap.

File: stl/inc/xutility [hunk 9]
```diff
@@ -4779,97 +4888,138 @@ _NODISCARD bool is_permutation(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2,
...
-    for (; _First1 != _Last1 && _First2 != _Last2; ++_First1, (void) ++_First2) {
+    for (;;) { // trim matching prefix
+        if (_First1 == _Last1) {
+            return _First2 == _Last2; // sequence equal
+        }
+
+        if (_First2 == _Last2) {
+            return false;
+        }
+
```
Comment (miscco) [inline stl/inc/xutility]: I think these two can be simplified to
```cpp
        if (_First2 == _Last2) {
            return _First1 == _Last1; // sequence equal
        }
```

File: stl/inc/xutility [hunk 10]
```diff
@@ -4779,97 +4888,138 @@ _NODISCARD bool is_permutation(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2,
...
                 return _Check_match_counts(_First1, _Last1, _First2, _Last2, _Pred);
-            } else {
-                return false; // lengths differ, fail
             }
+
+            return false; // sequence 1 shorter than sequence 2, fail
+        }
+
+        if (_Next2 == _Last2) {
+            return false; // sequence 1 longer than sequence 2, fail
```
Comment (miscco) [inline stl/inc/xutility]: Similarly we could shorten this to 
```cpp
if (_Next2 == _Last2) {
    return _Next1 == _Last1
        && _Check_match_counts(_First1, _Last1, _First2, _Last2, _Pred);
}
```

Or similary
```cpp
if (_Next2 == _Last2) {
    if (_Next1 == _Last1) {
        return _Check_match_counts(_First1, _Last1, _First2, _Last2, _Pred);
    }
    return false;
}
```

File: stl/inc/xutility (hunk 9 above)
Comment (BillyONeal) [reply stl/inc/xutility]: I don't think so?

image

Comment (miscco) [reply stl/inc/xutility]: Nah I missed my coffee

File: stl/inc/xutility (hunk 10 above)
Comment (miscco) [reply stl/inc/xutility]: Derp, coffee first...

File: stl/inc/algorithm (hunk 5 above)
Comment (BillyONeal) [reply stl/inc/algorithm]: There's no point to do that -- `__builtin_memcmp` is the same as `memcmp` at runtime, and the reinterpret_casts before already exclude this block from constexpr processing.

File: stl/inc/xutility (hunk 2 above)
Comment (BillyONeal) [reply stl/inc/xutility]: Because it's specifically the implementation that uses only the core language; the memmove implementation is also an implementation.

File: stl/inc/xutility (hunk 1 above)
Comment (BillyONeal) [reply stl/inc/xutility]: I still don't understand why you want the algorithm to do an extra increment it doesn't otherwise have to do, particularly when doing the one less increment makes some use cases work.

(But this is also off topic for this review :) )

File: stl/inc/algorithm (hunk 5 above)
Comment (CaseyCarter) [reply stl/inc/algorithm]: The point is throughput: we don't need to `if (is_constant_evaluated())` with `__builtin_memcmp` since it's valid at both compile time and at runtime. The body becomes:
```c++
if constexpr (decltype(_Equal_memcmp_is_safe(_First1, _First2, _Pred))::value) {
    return 0 == __builtin_memcmp(_First1, _First2, (_Last2 - _First2) * sizeof(*_First2));
} else {
    // as before
}
```

Comment (BillyONeal) [reply stl/inc/algorithm]: (1) I believe that violates the contract for `__builtin_memcmp` at constexpr time for us; I believe it accepts chars only. At least, we are guarding it that way in the only place we currently use it.

(2) That gives up on the optimization we're getting from the reinterpret_cast -- it'll introduce the divide + multiply from the pointer math.

(3) The throughput difference of parsing this loop is so So SO far in the noise I don't know why we're even worrying about it.

Comment (CaseyCarter) [reply stl/inc/algorithm]: (1) Apparently so: VSO-714690.

(2) "Divide + multiply" what? [Oh dear](https://godbolt.org/z/8poAr3). Nevermind.

(3) Size of source code is also a metric. It would be nice if we could make this function body a dozen lines shorter.

File: stl/inc/xutility (hunk 2 above)
Comment (CaseyCarter) [reply stl/inc/xutility]: It seems odd to me to name a library function on the basis that it doesn't call another library function - that fact doesn't seem significant to me. That said, I get that you want a name that contrasts with `_Copy_memmove`, and I have no better suggestion, so why not.

**Comment by BillyONeal** [comment]
FYI to reviewers, I'm going to wait till #423 merges to resolve comments here.

File: stl/inc/xutility [hunk 11]
```diff
@@ -3568,6 +3500,74 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
...
+};
+
+template <class _Source, class _Dest>
+struct _Ptr_copy_cat : _False_copy_cat {}; // return pointer copy optimization category for arbitrary iterators
+
+template <class _Source, class _Dest>
+struct _Ptr_copy_cat<_Source*, _Dest*>
+    : conditional_t<is_trivially_assignable_v<_Dest&, _Source&>,
+          _Ptr_cat_helper<remove_cv_t<_Source>, remove_cv_t<_Dest>>, _False_copy_cat> {
+}; // return pointer copy optimization category for pointers
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply stl/inc/xutility]: If they go away completely it doesn't matter where they are :)

File: stl/inc/algorithm (hunk 5 above)
Comment (BillyONeal) [reply stl/inc/algorithm]: RE: (2) @manbearian said they were looking at such problems once upon a time

File: stl/inc/xutility [hunk 12]
```diff
@@ -3568,6 +3500,74 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
 #endif // __cpp_lib_concepts
 
 // FUNCTION TEMPLATE copy
+template <class _Source, class _Dest>
+struct _Ptr_cat_helper { // determines _Ptr_cat's result in the most general case
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply stl/inc/xutility]: Hmm I think on the partial specializations they serve to motivate why the partial specialization is OK. I removed some of the obvious ones.

File: stl/inc/xutility [hunk 13]
```diff
@@ -3568,6 +3500,74 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
 #endif // __cpp_lib_concepts
 
 // FUNCTION TEMPLATE copy
+template <class _Source, class _Dest>
+struct _Ptr_cat_helper { // determines _Ptr_cat's result in the most general case
+    using _USource                        = _Unwrap_enum_t<_Source>;
+    using _UDest                          = _Unwrap_enum_t<_Dest>;
+    static constexpr bool _Really_trivial = conjunction_v<
+        bool_constant<sizeof(_USource) == sizeof(_UDest) && is_same_v<bool, _USource> == is_same_v<bool, _UDest>>,
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply stl/inc/xutility]: That's a huge change I would prefer to do in a separate PR. Filed https://github.com/microsoft/STL/issues/431

File: stl/inc/xutility (hunk 12 above)
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/xutility]: To me, "determines `_Ptr_cat`'s result" says that `_Ptr_cat_helper` implements `_Ptr_cat` - which is ~obvious - and each of the "when [condition]" bits simply repeat the pattern that the partial specialization matches:

* `_Ptr_cat_helper<_Elem, _Elem>`..."when the types are the same"

* `_Ptr_cat_helper<_Anything*, const _Anything*>`..."when all we do is add `const` to a pointer"

I don't think the comments provide any new information or justification.

File: stl/inc/yvals_core.h [hunk 14]
```diff
@@ -1022,7 +1022,8 @@
 #define __cpp_lib_unwrap_ref              201811L
```
Comment (StephanTLavavej) [inline stl/inc/yvals_core.h]: The newly-implemented features need to be listed in the comment at the top.

File: stl/inc/numeric [hunk 15]
```diff
@@ -75,25 +75,26 @@ _NODISCARD _CONSTEXPR20_ICE _Ty reduce(const _InIt _First, const _InIt _Last, _T
...
-    } else
-#endif // __cpp_lib_is_constant_evaluated
-    {
-        if constexpr (_Plus_on_arithmetic_ranges_reduction_v<_Unwrapped_t<const _InIt&>, _Ty, _BinOp>) {
-            (void) _Reduce_op; // TRANSITION, VSO-486357
-            return _Reduce_plus_arithmetic_ranges(_UFirst, _ULast, _Val);
-        } else {
+        if (_STD is_constant_evaluated()) {
             for (; _UFirst != _ULast; ++_UFirst) {
                 _Val = _Reduce_op(_STD move(_Val), *_UFirst); // Requirement missing from N4713
```
Comment (StephanTLavavej) [inline stl/inc/numeric]: This repeats the "classic implementation"; for #414 I hoped that we would be able to avoid that. Did you intend to choose another approach?

File: stl/inc/algorithm [hunk 16]
```diff
@@ -1243,7 +1271,7 @@ _NODISCARD _FwdIt1 find_first_of(_ExPo&& _Exec, const _FwdIt1 _First1, const _Fw
 
 // FUNCTION TEMPLATE swap_ranges
 template <class _FwdIt1, class _FwdIt2>
-_FwdIt2 _Swap_ranges_unchecked(_FwdIt1 _First1, const _FwdIt1 _Last1, _FwdIt2 _First2) {
+constexpr _FwdIt2 _Swap_ranges_unchecked(_FwdIt1 _First1, const _FwdIt1 _Last1, _FwdIt2 _First2) {
```
Comment (StephanTLavavej) [inline stl/inc/algorithm]: Why is this plain `constexpr`? This calls `iter_swap` which is `_CONSTEXPR20`.

File: stl/inc/algorithm [hunk 17]
```diff
@@ -2946,34 +2994,34 @@ _NODISCARD bool binary_search(_FwdIt _First, _FwdIt _Last, const _Ty& _Val, _Pr
...
 template <class _FwdIt, class _Ty>
-_NODISCARD bool binary_search(_FwdIt _First, _FwdIt _Last, const _Ty& _Val) {
+_NODISCARD _CONSTEXPR20 bool binary_search(_FwdIt _First, _FwdIt _Last, const _Ty& _Val) {
     // test if _Val equivalent to some element, using operator<
     return _STD binary_search(_First, _Last, _Val, less<>());
 }
 
 // FUNCTION TEMPLATE merge
-inline _Distance_unknown _Idl_dist_add(_Distance_unknown, _Distance_unknown) {
+constexpr _Distance_unknown _Idl_dist_add(_Distance_unknown, _Distance_unknown) {
```
Comment (StephanTLavavej) [inline stl/inc/algorithm]: Pre-existing: The `_Idl_dist_add` overloads should be `_NODISCARD`.

File: stl/inc/xutility [hunk 18]
```diff
@@ -4750,15 +4859,15 @@ bool _Check_match_counts(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2, _FwdI
...
+_NODISCARD _CONSTEXPR20 bool _Is_permutation_unchecked(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2, _Pr _Pred) {
     // test if [_First1, _Last1) == permuted [_First2, ...), using _Pred
-    for (; _First1 != _Last1; ++_First1, (void) ++_First2) {
+    for (; _First1 != _Last1; ++_First1, (void) ++_First2) { // trim matching prefix
         if (!_Pred(*_First1, *_First2)) {
-            // found first inequality, check match counts in suffix narrowing _Iter_diff_t<_FwdIt1> to
-            // _Iter_diff_t<_FwdIt2> is OK because if the 2nd range is shorter than the 1st, the user already
-            // triggered UB
+            // found first inequality, check match counts in suffix
+            //
```
Comment (StephanTLavavej) [inline stl/inc/xutility]: We generally don't have empty comment lines like this (there may be exceptions but I don't like 'em).

File: stl/inc/xutility [hunk 19]
```diff
@@ -4779,97 +4888,138 @@ _NODISCARD bool is_permutation(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2,
...
+_NODISCARD _CONSTEXPR20 bool is_permutation(
+    _FwdIt1 _First1, _FwdIt1 _Last1, _RightTy (&_First2)[_RightSize], _Pr _Pred) {
     // test if [_First1, _Last1) == permuted [_First2, ...), using _Pred
     return _STD is_permutation(_First1, _Last1, _Array_iterator<_RightTy, _RightSize>(_First2), _Pass_fn(_Pred));
 }
 #endif // _ITERATOR_DEBUG_ARRAY_OVERLOADS
 
 template <class _FwdIt1, class _FwdIt2>
-bool is_permutation(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2) {
+_CONSTEXPR20 bool is_permutation(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2) {
```
Comment (StephanTLavavej) [inline stl/inc/xutility]: Pre-existing: This Standard overload is egregiously and mysteriously missing `_NODISCARD`.

File: stl/inc/xutility (hunk 18 above)
Comment (StephanTLavavej) [reply stl/inc/xutility]: Disregard; this was part of the previous PR and I didn't complain then. The lines are no longer being edited here.

File: stl/inc/numeric (hunk 15 above)
Comment (BillyONeal) [reply stl/inc/numeric]: Nothing suggested in #414 avoids repeating the classic implementation, only moving it such only one copy of it is considered for a given setting of the if constexpr branch.

Comment (StephanTLavavej) [reply stl/inc/numeric]: I suggested avoiding the repetition. See:

> So perhaps the best pattern is:

```
if constexpr (_Plus_on_arithmetic_ranges_reduction_v) {
    if (!_STD is_constant_evaluated()) {
        return _Reduce_plus_arithmetic_ranges(ARGS);
    }
}

return CLASSIC_IMPLEMENTATION;
```

Comment (BillyONeal) [reply stl/inc/numeric]: I believe that pattern will trip DevCom-889321.

Comment (StephanTLavavej) [reply stl/inc/numeric]: The proposed pattern is `is_constant_evaluated` within `if constexpr`; your filed bug is about `if constexpr` within `is_constant_evaluated`.

Comment (BillyONeal) [reply stl/inc/numeric]: OTOH DevCom-889321 was the opposite (if constexpr inside is_constant_evaluated). Let me try this and see if it works out...

Comment (BillyONeal) [reply stl/inc/numeric]: Hmmm well it seems to work so I'm not going to look a gift horse in the mouth ¯\_(ツ)_/¯

File: stl/inc/algorithm (hunk 16 above)
Comment (BillyONeal) [reply stl/inc/algorithm]: I guess nobody cares about that IFNDR

File: stl/inc/numeric [hunk 20]
```diff
@@ -211,29 +206,23 @@ _NODISCARD _CONSTEXPR20_ICE _Ty transform_reduce(
...
+                      _BinOp1, _BinOp2>) {
 #ifdef __cpp_lib_is_constant_evaluated
-    // TRANSITION, DevCom-878972
-    if (_STD is_constant_evaluated()) {
-        for (; _UFirst1 != _ULast1; ++_UFirst1, (void) ++_UFirst2) {
-            _Val = _Reduce_op(_STD move(_Val), _Transform_op(*_UFirst1, *_UFirst2)); // Requirement missing from N4713
-        }
-        return _Val;
-    } else
+        // TRANSITION, DevCom-878972
```
Comment (StephanTLavavej) [inline stl/inc/numeric]: Is this still TRANSITION now that you've eliminated the code duplication?

File: stl/inc/xutility [hunk 21]
```diff
@@ -3663,27 +3662,38 @@ _DestTy* copy(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Last, _DestTy (&_Dest)[_DestSize
 #endif // _HAS_CXX17
 
 // FUNCTION TEMPLATE copy_n
+template <class _InIt, class _Diff, class _OutIt>
+constexpr _OutIt _Copy_n_core(_InIt _First, _Diff _Count, _OutIt _Dest) {}
```
Comment (StephanTLavavej) [inline stl/inc/xutility]: `_Copy_n_core` appears to be unused, and it has an empty function body.

Comment (BillyONeal) [reply stl/inc/xutility]: 🙄

File: stl/inc/yvals_core.h [hunk 22]
```diff
@@ -39,6 +39,7 @@
 // P0063R3 C11 Standard Library
 // P0074R0 owner_less<>
 // P0092R1 <chrono> floor(), ceil(), round(), abs()
+// P0202R3 constexpr For <algorithm> And exchange()
```
Comment (StephanTLavavej) [inline stl/inc/yvals_core.h]: These features are incorrectly listed in the "implemented unconditionally" section.

File: stl/inc/xutility [hunk 23]
```diff
@@ -4520,17 +4566,26 @@ bool _Lex_compare_unchecked(
...
-    const auto _Num1 = static_cast<size_t>(_Last1 - _First1);
-    const auto _Num2 = static_cast<size_t>(_Last2 - _First2);
-    const int _Ans   = _CSTD memcmp(_First1, _First2, _Num1 < _Num2 ? _Num1 : _Num2);
-    return _Memcmp_pr{}(_Ans, 0) || (_Ans == 0 && _Num1 < _Num2);
+#ifdef __cpp_lib_is_constant_evaluated
+    if (_STD is_constant_evaluated()) {
+        return _Lex_compare_unchecked(_First1, _Last1, _First2, _Last2, _Pred, _Lex_compare_optimize<void>{});
+    }
+#endif // __cpp_lib_is_constant_evaluated
+    {
```
Comment (StephanTLavavej) [inline stl/inc/xutility]: The braces here appear to be a relic.

File: stl/inc/xutility [hunk 24]
```diff
@@ -4636,7 +4679,26 @@ _InIt _Find_unchecked1(_InIt _First, const _InIt _Last, const _Ty& _Val, false_t
...
+    if (!_Within_limits(_First, _Val)) {
+        return _Last;
+    }
+
+#ifdef __cpp_lib_is_constant_evaluated
+    if (_STD is_constant_evaluated()) {
+        return _Find_unchecked1(_First, _Last, _Val, false_type{});
+    }
+#endif // __cpp_lib_is_constant_evaluated
+    {
```
Comment (StephanTLavavej) [inline stl/inc/xutility]: Ditto relic braces.

File: stl/inc/yvals_core.h (hunk 22 above)
Comment (BillyONeal) [reply stl/inc/yvals_core.h]: There are days when I need to kick myself. This is one of those days.

**Comment by StephanTLavavej** [comment]
Microsoft-internal MSVC-PR-224184 completed, so I went ahead and merged this.
